### DISCUSSION TITLE
Moving cable endpoints

### DISCUSSION THREAD

**Discussion by candlerb** [netbox-community/netbox#7792] [General]
I've just installed a new switch and moved about 2 dozen cables from the old one to the new one.  This time, I *really* want to avoid the tedium of deleting cables and individually creating new ones, where everything is identical to the old one apart from one endpoint.

I am wondering: if I use nbshell or a custom script to change termination_a or termination_b and save back to the database, will this  trigger all the necessary dependent changes?   Specifically I'm thinking of the CablePath, and the various columns on the dcim_interface table from CableTerminatino and PathEndpoint (`cable_id`, `_cable_peer_id`, `_cable_peer_type_id`, `_path_id`)

Or would running `manage.py trace_paths [--force]` fix all these?  From the [code](https://github.com/netbox-community/netbox/blob/develop/netbox/dcim/management/commands/trace_paths.py) I expect that it would fix CablePath but I'm not sure about the attributes on Interface.

**Comment by jeremystretch** [comment]
#4881 is open and just waiting for a volunteer.

**Reply by candlerb** [reply]
It would certainly be nice to have this in the UI.

But for now my question remains: if I simply modify (one of) the endpoints on a Cable model instance, will all necessary changes propagate? And if not, would `manage.py trace_paths` be sufficient to correct everything?

(Aside: I haven't tried PATCHing the endpoints on a cable via the REST API. But it [looks like](https://github.com/netbox-community/netbox/blob/v3.0.9/netbox/dcim/api/serializers.py#L755-L756) this won't be allowed)

**Comment by candlerb** [comment]
In the end I solved this via CSV export and import, although some work is required because the formats are very different (#5952).

I made a new export template (which also turned up issue #7813):

```
side_a_device,side_a_type,side_a_name,side_b_device,side_b_type,side_b_name,type,status,label,color,length,length_unit
{%- for c in queryset %}
{%- if c.termination_a_type.app_label != 'circuits' and c.termination_b_type.app_label != 'circuits' %}
{{ c.termination_a.device.name }},{{ c.termination_a_type.app_labeled_name|replace(" | ",".")|replace(" ","") }},{{ c.termination_a.name }},{{ c.termination_b.device.name }},{{ c.termination_b_type.app_labeled_name|replace(" | ",".")|replace(" ","") }},{{ c.termination_b.name }},{{ c.type }},{{ c.status }},"{{ c.label }}",{{c.color}},{{ c.length|replace("None","") }},{{ c.length_unit }}
{%- endif %}
{%- endfor %}
```

Then a manual search-replace to change the endpoint device name and the port number format; delete the old cables; import the new cables.

**Comment by candlerb** [comment]
I am shortly about to have to move another whole bunch of cables from one device to another.

I am considering writing a custom script for this.  But my question from before was unanswered:

> if I use nbshell or a custom script to change termination_a or termination_b and save back to the database, will this trigger all the necessary dependent changes?

That is, is there *any* defined way in Python code to change a single cable endpoint, or would this leave the database in an inconsistent state?  The model is too complex for me to understand all the potential knock-on effects.

**Reply by candlerb** [reply]
Ah, I think I found the answer in [code](https://github.com/netbox-community/netbox/blob/v3.1.4/netbox/dcim/models/cables.py#L180-L196):

```
    def clean(self):
...
        # If editing an existing Cable instance, check that neither termination has been modified.
        if self.pk:
            err_msg = 'Cable termination points may not be modified. Delete and recreate the cable instead.'
```

**Comment by grinapo** [comment]
I see all open issues were closed by the bot, so cannot comment there.
I see that simple editing was disabled due to problematic integrity keeping, and people are suggested to delete and recreate the cable.
This is suboptimal, as the physical cables are usually keeping all their attributes, and also usually one endpoint.

Would it be easier and possible to:
1. Provide an interface which would **delete** the cable, and **pre-fill a new cable with the data of the old** (just deleted) cable, which could be inserted? This would help individual cable moves. Or:
2. Make it possible to keep physical "dangling" cables, to be re-used in new connection? (← not so good solution, but would make it much simpler to pick one from the unused cables, which already have their properties set.)
